### ISSUE TITLE
Screen readers and keyboard navigation of the new main menu

### ISSUE THREAD

**Issue by ognil** [wet-boew/wet-boew#8556]
Hi guys,

Please visit canada.ca and try to use NVDA or JAWS 18 screen readers.
I use Chrome, NVDA and JAWS 18 screen readers.
Without a screen reader, it works OK but with a screen reader the keyboard navigation (arrow keys) of the main menu stop working.

George

**Comment by duboisp** [comment]
Thanks @ognil for reporting that issue, I am looking at it.

**Comment by duboisp** [comment]
I was able to reproduce the issue with NVDA in IE11. It works fine in Firefox.

**Comment by thekodester** [comment]
Could it be related to the removal of `aria-posinset `and `aria-setsize`, and the introduction of roles like `role="presentation"`?

**Comment by duboisp** [comment]
@thekodester no it's not about ```aria-posinset``` and ```aria-setsize```.

The issue seems to be related the screen reader mode don't change to "Focus mode" but remain in "Browsing mode" when the button receive the focus. More info here: https://github.com/nvaccess/nvda/issues/7898

Adding the ```role="application"``` at the main menu container fix the issue but I am not sure if that is the best solution. We will need to do more test with various browser and OS.

**Comment by duboisp** [comment]
@ognil are you able to do a quick test with the test page bellow by using Chrome with a screen reader and let me know if you can operate the main menu? It contains a fix that resolved the issue I had with IE11 and NVDA.

Test page: https://universallabs.org/wet/labs/gcwebmenu/content-en.html

**Comment by ognil** [comment]
Thank you Pierre,

 

Unfortunately it still  doesn’t work with  Google Chrome Version 71.0.3578.98 (Official Build) (64-bit) with JAWS v.18 and NDVA. 

 

I tested this page:

https://universallabs.org/wet/labs/gcwebmenu/content-en.html

 

George

**Comment by duboisp** [comment]
@ognil what is the issue you have? Is the menu are still keyboard operable with the arrow key? How you are able to operate them?

I did some test and here my result:

| Browser | Screen reader | Date | Test page | Results |
| ---- | ---- | ---- | ---- |  ---- |
| Chrome | (none) |  2019-02-19 | Canada.ca | OK, The menu are keyboard operable as expected |
| Chrome | JAWS |  2019-02-19 | Canada.ca | OK, The menu are keyboard operable as expected |
| Chrome | NVDA | 2019-02-19 | Canada.ca | Buggy, The menu are keyboard operable but the user need to go through all the menu item. Down and up arrow work, but not right and left arrow |
| Chrome | (none) | 2019-02-19 | Test page with the fix | OK, The menu are keyboard operable as expected |
| Chrome | JAWS | 2019-02-19 | Test page with the fix | OK, The menu are keyboard operable as expected |
| Chrome | NVDA | 2019-02-19 | Test page with the fix | OK, The menu are keyboard operable as expected |
| Edge | (none) | 2019-02-19 | Canada.ca | OK, The menu are keyboard operable as expected  |
| Edge | JAWS | 2019-02-19 | Canada.ca | OK, The menu are keyboard operable as expected |
| Edge | NVDA | 2019-02-19 | Canada.ca | Buggy, The menu are keyboard operable but the user need to go through all the menu item. Down and up arrow work, but not right and left arrow |
| Edge | (none) | 2019-02-19 | Test page with the fix | OK, The menu are keyboard operable as expected |
| Edge | JAWS | 2019-02-19 | Test page with the fix | OK, The menu are keyboard operable as expected |
| Edge | NVDA| 2019-02-19 | Test page with the fix | Buggy, The menu are keyboard operable but the user need to go through all the menu item. Down and up arrow work, but not right and left arrow |

**Comment by duboisp** [comment]
I released GCWeb v5.0.1 today which contains a fix that had improved the user experience. The same fix are going to be applied on canada.ca during the following days.

We have done some additional <a href="https://wet-boew.github.io/themes-dist/GCWeb/gcweb-theme/test-menu.html#WAWG-test-result">testing of the menu with screen reader</a>. A few usability issue has been reported and they confirmed the fix improved the navigability of the menu for some screen reader and browser combination.

That testing revealed that adding instruction on how to navigate the menu would be an important improvement for screen reader user. So I will am going to keep this issue open until that would be completed.

@ognil if our testing don't represent to your experience please let us know with some additional details like how we can reproduce it, what behaviour you are experiencing...

**Comment by nricherbrule** [comment]
@bsouster After significant testing in Firefox, Chrome and IE with NVDA, the keyboard-only navigation does not seem impacted in any way. This ticket seems to be resolved and can be closed.
